Tickets/dm 30343 (#2)
lsst-ts/ts_M1M3ThermalFPGA: README.md

14. Copy resulting lvbitx file to ts_m1m3thermal/Bitfiles, and NiFpga_M1M3ThermalFPGA.h to ts_m1m3Thermal/src/NiFpga
15. Recompile ts_M1M3Thermal

## Overview

The FPGA design makes heavy use of FIFOs and **S**ingle **C**ycle **T**imed
**L**oop (**SCTL**) for critical hardware loops. As FPGA has to sample DIOs
(serial lines, pumps, flow meters,.. ), it's critical those are properly timed
and running all the time. In a classic front/back processing design (see e.g.
Linux Kernel interrupt handling), sampling code just record the values or bits
and bytes and ship them to a FIFO queues for handling. SCTLs, which are
guarantee to take one clock tick, are ideal for handling inputs and outputs.

Receiving (reading) code works by placing values into FIFO. Transmitting
(writing) code works by reading (without timeout) from FIFO, and if something
is available, act accordingly. This is coupled with DIO states (e.g. to make
sure next bit/byte on serial line is transmitted after the current).

## Command multiplexing

Commands, followed by arguments are filled into _Software
Resources/CommandFIFO_. _Commands/CommandFIFOMultiplexer_ read this and

> multiplex the command to various handlers _(This is exactly how CPUs are

Reviewer: ```suggestion
multiplex the command to various handlers _(This is exactly how CPUs

```

> handling instructions from binary code)_. Handlers fills in queues (e.g. ModBus

Reviewer: ```suggestion
handle instructions from binary code)_. Handlers fills in queues (e.g. ModBus

```

> writes queue with instructions to write if you are writing to a queue).

Reviewer: ```suggestion
creates a queue with instructions to write if you are pushing to a queue).

```

Author: This is now the full paragraph:

```
Commands, followed by arguments are filled into _Software
Resources/CommandFIFO_. _Commands/CommandFIFOMultiplexer_ read this and
multiplex the command to various handlers _(This is exactly how CPUs handle
instructions from binary code)_. Handlers fills in queues (e.g. ModBus transmit
command pushes bytes to be written into a queue associated with the given bus).
**SCTL**s are handling low level IO. All this resembles a hot potato handling - 
the FPGA is pushing data through various FIFOs to make sure all incoming 
request are handled and output buffers are filled with responses.
```
